### PR TITLE
feat(examples): add code review agent with skills, sandbox and sqlite storage

### PR DESCRIPTION
**Pull request by Yukua1**
## What changed

Add `examples/code_review_agent`, an automatic code review agent prototype for #2004
that combines Skills, sandbox execution, and database-backed audit storage:

- **CLI**: accepts `--diff-file`, `--repo-path` (git working tree diff), `--files`,
  or `--fixture`; writes `review_report.json` / `review_report.md` and persists
  everything to SQLite (queryable via `--task-id`).
- **code-review Skill**: `SKILL.md`, rule catalog (`docs/rules.md`), and sandbox
  scripts (`diff_summary.sh`, `secret_scan.sh`, `go_static_checks.sh`). Skill
  scripts run through the framework `tool/skill` tools (`skill_load`/`skill_run`).
- **Rules**: 8 deterministic Go-focused detectors covering security, goroutine/context
  leaks, resource lifecycle, error handling, database transaction lifecycle, missing
  tests, and reliability — above the 4-category requirement.
- **Sandbox**: `managed` (default, `codeexecutor/sandbox`), `container`, and `e2b`
  via the shared `codeexecutor.Engine`; `local-dev` is an explicit development
  fallback, `mock`/`--dry-run` supports offline runs. Optional `staticcheck`.
- **Governance**: every external command is gated by a `tool.PermissionPolicy`
  implementation; only Go checks and skill scripts are allow-listed, high-risk
  commands are denied or routed to human review, and all decisions are persisted.
- **Model modes**: `rule-only` (default), `fake-model` (full `llmagent` + `runner`
  chain with a deterministic offline model, no API key), and `llm`
  (OpenAI-compatible). Only redacted diffs reach the model; model failures degrade
  to rule-only results and are recorded as `model_error` exceptions.
- **Noise control**: findings deduplicated by file/line/rule/category keeping the
  highest severity/confidence; confidence buckets split into findings /
  needs_human_review / warnings; every keep/demote/drop is persisted as a
  filter decision.
- **Storage**: SQLite schema (tasks, findings, sandbox runs, permission decisions,
  filter decisions, reports, artifacts) behind a swappable `store.Store` interface.
- **Safety boundary**: per-command timeouts, 4 KiB output caps, env allow-list,
  redaction before model/report/persistence, 2 MiB artifact cap, failure records.
- **Monitoring**: metrics summary (total/sandbox duration, tool calls, permission
  denies, severity/exception/filter distributions) plus optional OTLP trace export
  via `--telemetry-endpoint`.
- **Samples**: 9 fixtures (clean, secret leak, goroutine/context leak, unclosed
  resource, DB lifecycle, missing test, duplicate findings, sandbox failure,
  redaction) with curated expected outputs, README, and DESIGN.md / DESIGN.zh_CN.md.

## Why

Implements #2004. The issue asks for a verifiable review pipeline rather than
"LLM comments on code": deterministic rule scanning first, sandboxed checks behind
permission gating, structured findings, noise control, and a complete audit trail
in persistent storage. Everything is example-scoped; no library APIs change.

## Testing

- `go build ./... && go test ./...` in `examples/code_review_agent` — all packages pass.
- Fixture e2e: all 9 fixtures produce reports; expected rule IDs asserted; curated
  outputs under `testdata/expected` are re-verified against real runs
  (`TestExpectedOutputsStayInSync`).
- Redaction: reports and database rows are asserted to never contain the fixture
  secrets (`assertNoFixtureSecrets`).
- `fake-model` mode covers the whole agent chain (prompt, streaming, JSON parsing,
  merge, persistence) without an API key; `llm` failure degrades to rule-only.
- Sandbox failure/timeout paths do not crash the review and are counted as
  exceptions (`TestSandboxFailureFixtureDegradesGracefully`, sandboxrunner tests).
- `gofmt -l .` and `go vet ./...` are clean. Container/e2b paths are compile- and
  unit-tested only (need Docker / E2B credentials).

## Notes for reviewers

- Example-only change under `examples/`; no public API impact.
- Rules are deterministic pattern matchers: `GOR001` matches `go func(...)` forms
  only and resource/secret detectors use fixed needle lists, so hidden-sample
  recall depends on pattern coverage — feedback on additional patterns welcome.
- Default `--sandbox managed` uses the OS sandbox; `container` needs Docker and
  `e2b` needs `E2B_API_KEY`.
- Redaction runs both per-field and once more over the rendered markdown as
  defense-in-depth; the double pass is intentional.

Refs #2004

### CODE REVIEW

File: examples/code_review_agent/sandboxrunner/runner.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		Cmd:      parts[0],
+		Args:     parts[1:],
+		Cwd:      codeexecutor.DirWork,
+		Env:      sandboxEnv(cfg),
+		Timeout:  cfg.Timeout,
+		CleanEnv: true,
+	})
+	run := review.SandboxRun{
+		Command:       command,
+		Status:        "completed",
```
Comment (Flash-LHR) [inline examples/code_review_agent/sandboxrunner/runner.go:218]: Non-zero exit codes are reported as completed here, so failing checks are omitted from exception metrics. Set failed status when `res.ExitCode != 0`.
<details><summary>中文</summary>
这里把非零退出码记录为 completed，导致失败检查不进入异常指标。请在 `res.ExitCode != 0` 时将状态设为 failed。
</details>

File: examples/code_review_agent/skillrunner/skillrunner.go [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+	if err != nil {
+		return failedRun(spec.command, start, err)
+	}
+	var rr runResult
+	if err := json.Unmarshal(raw, &rr); err != nil {
+		return failedRun(spec.command, start, err)
+	}
+	run := review.SandboxRun{
+		Command:       spec.command,
+		Status:        "completed",
```
Comment (Flash-LHR) [inline examples/code_review_agent/skillrunner/skillrunner.go:356]: `skill_run` non-zero exits are recorded as completed, so failed scripts disappear from exception metrics. Set failed status when `rr.ExitCode != 0`.
<details><summary>中文</summary>
`skill_run` 非零退出码被记录为 completed，导致失败脚本不进入异常指标。请在 `rr.ExitCode != 0` 时将状态设为 failed。
</details>

**Comment by Yukua1** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Yukua1** [comment]
@coderabbitai review

File: examples/code_review_agent/main.go [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+
+// resolveSkillsRoot locates the bundled skills directory when
+// --skills-root is not set.
+func resolveSkillsRoot(root string) string {
+	if root != "" {
+		return root
+	}
+	return firstExistingPath(
+		filepath.Join("code_review_agent", "skills"),
+		"skills",
```
Comment (Rememorio) [inline examples/code_review_agent/main.go]: **P1: Do not let the reviewed repository shadow bundled skills**

The default lookup checks `skills` in the current working directory before the bundled example directory. If the binary is launched from a repository containing `skills/code-review`, that repository can replace the three scripts even though the user did not pass `--skills-root`; the path-based permission policy then treats those attacker-controlled scripts as allow-listed. This is especially dangerous with `local-dev`, where they execute on the host.

Resolve an immutable bundled or executable-relative skill directory by default, and only trust a working-directory path when the user explicitly supplies `--skills-root`. Add a test that changes into a directory with a forged `skills/code-review` tree and verifies it cannot shadow the bundled skill.

<details><summary>中文</summary>

**不要让被审仓库覆盖内置 Skill**

默认查找顺序会先检查当前工作目录下的 `skills`，再检查示例自带目录。如果从一个包含 `skills/code-review` 的被审仓库中启动二进制，即使用户没有传入 `--skills-root`，仓库也能替换这三个脚本；随后基于路径的权限策略还会把这些仓库控制的脚本当作已允许命令。使用 `local-dev` 时，它们会直接在宿主机执行，风险更高。

默认路径应指向不可被工作目录覆盖的内置目录或可执行文件相对目录；只有用户明确指定 `--skills-root` 时，才应信任外部路径。建议增加测试：切换到带伪造 `skills/code-review` 的目录后，确认它无法覆盖内置 Skill。

</details>

File: examples/code_review_agent/reviewagent/agent.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+	mdl, err := buildModel(cfg)
+	if err != nil {
+		return Output{}, err
+	}
+	if cfg.Timeout <= 0 {
+		cfg.Timeout = 60 * time.Second
+	}
+	start := time.Now()
+	out := Output{ModelCalls: 1}
+	content, err := invoke(ctx, cfg, mdl, BuildPrompt(files))
```
Comment (Rememorio) [inline examples/code_review_agent/reviewagent/agent.go]: **P1: Enforce redaction at the remote-model boundary**

`Review` sends `BuildPrompt(files)` directly to the selected model. The CLI currently redacts line bodies before calling this exported API, but `Review` itself documents no such precondition, and the CLI sanitizer leaves fields such as `NewPath` unchanged. A raw caller—or a secret-bearing filename in the CLI—can therefore send sensitive data to the OpenAI-compatible endpoint despite the stated redaction guarantee.

Apply comprehensive redaction inside `Review` immediately before building the prompt, covering every field included in the prompt. Add a boundary test with secrets in both line content and a file path so this invariant cannot be bypassed by another caller.

<details><summary>中文</summary>

**在远程模型调用边界强制脱敏**

`Review` 会直接把 `BuildPrompt(files)` 的结果发送给所选模型。当前 CLI 虽然会先处理代码行内容，但这个导出的 API 并没有声明调用者必须预先脱敏，而且 CLI 的处理也没有覆盖 `NewPath` 等字段。因此，其他调用者传入原始 `ChangedFile`，或文件名本身包含敏感信息时，数据仍可能被发送到 OpenAI 兼容端点，与“只发送脱敏 diff”的约束不符。

建议在 `Review` 内部、构造提示词之前统一完成完整脱敏，覆盖提示词会使用的所有字段。再增加一个边界测试，同时在代码行和文件路径中放入测试密钥，确保任何调用入口都无法绕过这条安全约束。

</details>

File: examples/code_review_agent/reviewagent/agent.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+		fmt.Fprintf(&b, "FILE: %s", file.NewPath)
+		if file.PackageName != "" {
+			fmt.Fprintf(&b, " (package %s)", file.PackageName)
+		}
+		b.WriteByte('\n')
+		for _, hunk := range file.Hunks {
+			for _, line := range hunk.Lines {
+				switch line.Kind {
+				case "added":
+					fmt.Fprintf(&b, "+ %d: %s\n", line.NewLine, line.Content)
```
Comment (Rememorio) [inline examples/code_review_agent/reviewagent/agent.go]: **P1: Apply the prompt cap before appending each line**

The size check happens only after `fmt.Fprintf` has appended the complete line. `ParseUnifiedDiff` accepts individual lines approaching 2 MiB, and `BuildPrompt` is exported, so one long added line can make the supposedly 30,000-byte prompt orders of magnitude larger before truncation is noticed. In LLM mode this can cause excessive request cost, model rejection, and silent degradation to rule-only results.

Compute the remaining budget before writing each line and append at most that amount, including the truncation marker in the limit. Add a test with a megabyte-sized line that asserts the returned prompt never exceeds the documented cap.

<details><summary>中文</summary>

**在追加每行内容之前执行提示词上限**

当前是在 `fmt.Fprintf` 已经写入整行之后才检查大小。`ParseUnifiedDiff` 允许接近 2 MiB 的单行输入，而 `BuildPrompt` 还是导出函数，因此只需一条超长新增行，就能让标称 30,000 字节的提示词在发现超限前膨胀几十倍。LLM 模式下这会带来额外调用成本、模型拒绝请求，并最终静默降级为纯规则结果。

应在写入每一行之前计算剩余容量，只追加允许的部分，并把截断标记也计入上限。建议增加一个包含兆字节级单行内容的测试，明确断言返回提示词不会超过约定大小。

</details>

File: examples/code_review_agent/sandboxrunner/runner.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+// runEngine stages the repo into a workspace and runs the command on the engine.
+func runEngine(ctx context.Context, eng codeexecutor.Engine, cfg Config, command string, start time.Time) review.SandboxRun {
+	ws, err := eng.Manager().CreateWorkspace(ctx, cfg.TaskID, codeexecutor.WorkspacePolicy{})
+	if err != nil {
+		return failedRun(command, start, err)
+	}
+	defer eng.Manager().Cleanup(ctx, ws)
+	if err := eng.FS().StageDirectory(ctx, ws, cfg.RepoPath, codeexecutor.DirWork, codeexecutor.StageOptions{}); err != nil {
```
Comment (Rememorio) [inline examples/code_review_agent/sandboxrunner/runner.go:199]: **P1: Provide an offline dependency strategy for sandboxed Go checks**

Only the repository is staged before `go test ./...` and `go vet ./...` run. The managed backend restricts networking and gives the process a fresh workspace `HOME`; the container backend also disables networking and points `GOPATH`/`GOCACHE` at empty `/tmp` directories. Consequently, any ordinary non-vendored module with third-party dependencies fails dependency resolution before either check can analyze the change, including under the default managed mode.

Please define an offline dependency policy: for example, stage a read-only module cache plus a writable build cache, require and detect vendoring, or classify unavailable dependencies as an explicit unsupported/skipped condition. A backend test using a small module with one external dependency would validate that the default path performs real checks without weakening network isolation.

<details><summary>中文</summary>

**为沙箱中的 Go 检查提供离线依赖方案**

在运行 `go test ./...` 和 `go vet ./...` 前，目前只把仓库内容放入工作区。managed 后端禁止网络访问，并为进程提供全新的工作区 `HOME`；container 后端同样关闭网络，还把 `GOPATH`、`GOCACHE` 指向空的 `/tmp` 目录。因此，任何未提交 vendor 且包含第三方依赖的普通 Go 模块，都会在真正分析代码之前就因为无法解析依赖而失败，默认 managed 模式也会受影响。

需要明确一套离线依赖策略，例如挂载只读模块缓存并提供可写构建缓存、要求并检测 vendor，或者把依赖不可用明确归类为 unsupported/skipped，而不是普通检查失败。建议增加一个只依赖单个外部模块的小型后端测试，确认默认路径能在不放开网络边界的情况下执行真实检查。

</details>

File: examples/go.mod [hunk 7]
```diff
@@ -1,9 +1,10 @@
 module trpc.group/trpc-go/trpc-agent-go/examples
 
-go 1.24.4
+go 1.25.0
```
Comment (Rememorio) [inline examples/go.mod]: **P1: Preserve the examples module's Go 1.24 compatibility**

This changes the minimum toolchain for every package in the shared `examples` module from Go 1.24.4 to Go 1.25.0, not just for the new example. Repository CI still configures `EXAMPLES_GO_VERSION` as 1.24.6, and existing example documentation promises Go 1.24.4 compatibility; users with toolchain auto-download disabled can no longer build any of these examples.

The added code and container submodule do not establish a need for Go 1.25. Please restore the previous directive and tidy/pin the new dependencies with that toolchain, or update CI and the compatibility documentation as an explicit repository-wide change.

<details><summary>中文</summary>

**保留 examples 模块的 Go 1.24 兼容性**

这里把共享 `examples` 模块的最低 Go 版本从 1.24.4 提升到了 1.25.0，影响的不只是新示例，而是该模块下的所有示例。仓库 CI 仍使用 `EXAMPLES_GO_VERSION=1.24.6`，已有示例文档也声明兼容 Go 1.24.4；关闭自动下载 toolchain 的用户将无法再构建这些示例。

当前新增代码以及 container 子模块都没有体现出必须使用 Go 1.25 的理由。建议恢复原来的版本，并用该版本整理、固定新增依赖；如果确实要整体升级，则应作为明确的仓库级兼容性变更同步更新 CI 和文档。

</details>

File: examples/code_review_agent/diffparser/parser.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+	flushHunk := func() {
+		if current != nil && currentHunk != nil {
+			current.Hunks = append(current.Hunks, *currentHunk)
+			currentHunk = nil
+		}
+	}
+	flushFile := func() {
+		flushHunk()
+		if current != nil && current.NewPath != "" {
```
Comment (Rememorio) [inline examples/code_review_agent/diffparser/parser.go]: **P2: Retain files deleted by the diff**

For a deletion, the `+++ /dev/null` header clears `NewPath`, so this condition discards the entire `ChangedFile` during flush. A deletion-only change is therefore parsed as if it touched no files: it disappears from the report and model context, and downstream audit logic cannot distinguish it from an empty diff.

Keep entries when either old or new path is present, and define an effective display/language path for deletions while preserving their deleted status. Cover both git-style and plain unified deletions in parser tests.

<details><summary>中文</summary>

**保留 diff 中被删除的文件**

删除文件时，`+++ /dev/null` 会把 `NewPath` 清空，而这里又要求 `NewPath` 非空才保存，导致整个 `ChangedFile` 被丢弃。只包含删除操作的变更会被解析成“没有修改任何文件”，既不会出现在报告和模型上下文中，后续审计也无法把它与空 diff 区分开。

建议在旧路径或新路径任一存在时都保留该条目，同时为删除文件明确展示路径、语言识别和删除状态的语义。解析器测试应覆盖 git 风格和普通 unified diff 的删除场景。

</details>

File: examples/code_review_agent/rules/rules.go [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+		if trimmed == "" || strings.HasPrefix(trimmed, "//") {
+			continue
+		}
+		add := func(f review.Finding) {
+			f.File = file.NewPath
+			f.Line = line.NewLine
+			f.Evidence = redaction.RedactText(strings.TrimSpace(f.Evidence))
+			findings = append(findings, f)
+		}
+		if secretLineRE.MatchString(trimmed) || strings.Contains(trimmed, "sk-") || strings.Contains(trimmed, "ghp_") {
```
Comment (Rememorio) [inline examples/code_review_agent/rules/rules.go:98]: **P2: Do not classify secret retrieval as a hard-coded credential**

This condition treats the presence of a secret-related identifier assignment as proof of a hard-coded secret, regardless of the right-hand side. Safe patterns such as `token := os.Getenv("TOKEN")` or `password := cfg.Password` therefore produce a critical `SEC001` finding with 0.96 confidence, which defeats the high-confidence noise-control contract.

Require evidence of a literal credential value before assigning critical confidence, or route identifier-only matches to a lower-confidence heuristic. Add negative tests for environment/configuration retrieval alongside the existing literal-secret fixture.

<details><summary>中文</summary>

**不要把安全读取密钥误判为硬编码凭据**

这里仅凭“与密钥相关的标识符发生赋值”就认定存在硬编码密钥，没有检查赋值右侧。像 `token := os.Getenv("TOKEN")`、`password := cfg.Password` 这类正确做法也会触发置信度 0.96 的 critical `SEC001`，直接破坏高置信度降噪的可信度。

只有检测到字面量凭据值时才应给出 critical 高置信度结论；仅命中变量名的情况可以降为低置信度启发式提示。建议在现有硬编码密钥用例之外，增加环境变量和配置读取的反向测试。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 7 inline comments below. The comments focus on issues worth addressing before merge.

Fresh review found seven actionable issues: the shared examples module now requires Go 1.25, model redaction and prompt-size limits are not enforced at their sinks, default skill discovery can execute repository-controlled scripts, offline sandboxes cannot resolve normal module dependencies, deleted files are dropped, and the secret rule emits high-confidence false positives. No previous review comments were provided. Recommended validation includes Go 1.24.6 compatibility, adversarial skill-root and prompt tests, a managed-sandbox module with an external dependency, deletion parsing, and safe secret-loading patterns.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 7 条行内评论。评论聚焦在合并前值得处理的问题。

本次审查发现 7 个可立即处理的问题：共享 examples 模块被整体提升到 Go 1.25；模型调用边界没有自行保证脱敏，提示词大小上限也可被单行输入突破；默认 Skill 查找可能执行被审仓库提供的脚本；离线沙箱无法解析常见的 Go 模块依赖；删除文件会被解析器丢弃；密钥规则还会把安全的环境变量读取误报为高置信度泄密。没有需要跟进的历史评论。建议重点验证 Go 1.24.6 兼容性、恶意工作目录下的 Skill 覆盖、超长及含敏感信息的提示词、带外部依赖的 managed 沙箱、删除文件 diff，以及安全读取密钥的场景。

</details>

File: examples/code_review_agent/main.go (hunk 3 above)
Comment (Rememorio) [reply examples/code_review_agent/main.go]: Fixed. The default now resolves only from the bundled source or executable directory, and the forged-CWD test covers the shadowing case.

<details><summary>中文</summary>

已修复。默认路径现在只从内置源码目录或可执行文件目录解析，伪造工作目录的测试也覆盖了覆盖攻击场景。

</details>

File: examples/code_review_agent/reviewagent/agent.go (hunk 4 above)
Comment (Rememorio) [reply examples/code_review_agent/reviewagent/agent.go]: Fixed. `Review` now enforces redaction immediately before prompt construction, including paths, package names, and line content, with boundary coverage for both path and content secrets.

<details><summary>中文</summary>

已修复。`Review` 现在会在构造提示词前统一脱敏路径、包名和代码行内容，并通过边界测试同时覆盖路径与内容中的密钥。

</details>

File: examples/code_review_agent/reviewagent/agent.go (hunk 5 above)
Comment (Rememorio) [reply examples/code_review_agent/reviewagent/agent.go]: Fixed. Each chunk is bounded before writing, the truncation marker is included in the cap, and the oversized-line test asserts the result never exceeds 30,000 bytes.

<details><summary>中文</summary>

已修复。每个内容块都会在写入前受剩余预算限制，截断标记也计入上限，超长单行测试明确断言结果不超过 30,000 字节。

</details>

File: examples/code_review_agent/sandboxrunner/runner.go (hunk 6 above)
Comment (Rememorio) [reply examples/code_review_agent/sandboxrunner/runner.go:199]: The direct `RunChecks` path is fixed, but `skillrunner.buildScripts` still launches `go_static_checks.sh` against the staged repository with an empty offline cache and no matching vendor/skip decision. Please apply the same offline policy to that path (or remove the duplicate Go-check path) and cover the full review result for an unvendored external dependency.

<details><summary>中文</summary>

直接 `RunChecks` 链路已经修复，但 `skillrunner.buildScripts` 仍会对暂存仓库执行 `go_static_checks.sh`，使用空的离线缓存且没有对应的 vendor/skip 判断。请将同一离线策略应用到该链路，或移除重复的 Go 检查，并补充包含未 vendor 外部依赖的完整评审测试。

</details>

File: examples/go.mod (hunk 7 above)
Comment (Rememorio) [reply examples/go.mod]: Fixed. The shared examples module is back on Go 1.24.4, preserving the existing compatibility baseline.

<details><summary>中文</summary>

已修复。共享 examples 模块已恢复为 Go 1.24.4，保留了原有兼容性基线。

</details>

File: examples/code_review_agent/diffparser/parser.go (hunk 8 above)
Comment (Rememorio) [reply examples/code_review_agent/diffparser/parser.go]: Fixed. Deletions are retained using the old path, marked with `Deleted`, and both git-style and plain unified forms are covered by parser tests.

<details><summary>中文</summary>

已修复。删除文件会使用旧路径保留并设置 `Deleted`，解析测试也覆盖了 Git 风格和普通 unified diff 两种形式。

</details>

File: examples/code_review_agent/rules/rules.go (hunk 9 above)
Comment (Rememorio) [reply examples/code_review_agent/rules/rules.go:98]: Fixed. Critical confidence now requires literal credential evidence, while environment/config retrievals are routed to the low-confidence warning bucket with negative coverage.

<details><summary>中文</summary>

已修复。现在只有字面量凭据证据才会触发 critical 高置信度结论，环境变量和配置读取会进入低置信度 warning，并已有反向测试覆盖。

</details>

File: examples/code_review_agent/redaction/redaction.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+// trpc-agent-go is licensed under the Apache License Version 2.0.
+//
+
+// Package redaction removes secrets before data is persisted or reported.
+package redaction
+
+import "regexp"
+
+var patterns = []*regexp.Regexp{
+	regexp.MustCompile(`(?i)([A-Za-z_][A-Za-z0-9_]*(?:api[_-]?key|token|password|passwd|secret)[A-Za-z0-9_]*)(\s*[:=]\s*)(?:\\?["'])?[^\\"',;\s]+(?:\\?["'])?`),
```
Comment (Rememorio) [inline examples/code_review_agent/redaction/redaction.go:23]: **P1: Redact the complete quoted secret value**

The assignment regex stops at whitespace even when the value is quoted. For example, `password = "correct horse battery staple"` becomes `password = [REDACTED_SECRET] horse battery staple"`, leaving most of the credential intact.

Because `Review` uses this function at the remote-model boundary, the remainder is sent to the model and can also survive report and database redaction. Handle quoted values through their closing quote separately from unquoted tokens, and add a boundary test with a multiword passphrase.

<details><summary>中文</summary>

**完整脱敏引号内的密钥值**

赋值正则即使遇到引号值，也会在第一个空格处停止。例如 `password = "correct horse battery staple"` 只会隐藏 `correct`，后面的 `horse battery staple` 仍然保留。

`Review` 在远程模型边界直接依赖这个函数，因此剩余内容会被发送给模型，也可能继续进入报告和数据库。建议分别处理引号值与非引号值，确保引号内的完整内容都被替换，并增加包含多词口令的边界测试。

</details>

File: examples/code_review_agent/report/report.go [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+// redactReport applies secret redaction to every report field.
+func redactReport(r review.ReviewReport) review.ReviewReport {
+	out := r
+	out.Task.InputSummary = redaction.RedactText(out.Task.InputSummary)
+	out.Task.Error = redaction.RedactText(out.Task.Error)
+	out.Summary = redaction.RedactText(out.Summary)
+	out.Files = make([]review.ChangedFile, len(r.Files))
+	copy(out.Files, r.Files)
+	for i := range out.Files {
```
Comment (Rememorio) [inline examples/code_review_agent/report/report.go:160]: **P1: Redact path-bearing fields before reporting and persistence**

`redactReport` copies each `ChangedFile` unchanged and only redacts line content. `Task.RepoPath`, `OldPath`/`NewPath`, `PackageName`, `Finding.File`, and `FilterDecision.File` therefore remain raw in the JSON report; `SaveFindings` and `SaveFilterDecisions` also insert their file fields without redaction.

A secret-bearing path such as the AWS-key filename used by the new model-boundary test will still leak into artifacts and SQLite. Apply one complete structural redactor to every path-bearing field at both boundaries, and add an end-to-end test that inspects the JSON and raw database rows.

<details><summary>中文</summary>

**在报告和持久化前脱敏路径字段**

`redactReport` 会原样复制 `ChangedFile`，只处理代码行内容。因此 `Task.RepoPath`、`OldPath`/`NewPath`、`PackageName`、`Finding.File` 和 `FilterDecision.File` 仍会以原值进入 JSON；`SaveFindings` 与 `SaveFilterDecisions` 写库时也没有脱敏文件字段。

像新增模型边界测试中的 AWS Key 文件名，仍会泄漏到产物和 SQLite。建议使用同一套完整的结构化脱敏逻辑覆盖所有路径字段，并增加端到端测试，直接检查 JSON 和数据库原始行。

</details>

File: examples/code_review_agent/main.go [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+	// fail finalizes the task so persistence errors never leave it in
+	// the "running" state without an audit trail.
+	fail := func(err error) error {
+		task.Status = review.StatusFailed
+		task.Error = err.Error()
+		_ = db.FinishTask(ctx, task)
+		return err
+	}
+
+	artifacts, err := reportwriter.Write(cfg.outDir, report)
```
Comment (Rememorio) [inline examples/code_review_agent/main.go:307]: **P2: Do not publish a completed report before persistence succeeds**

The report is written with `StatusCompleted` before any findings, audit decisions, artifacts, or report metadata are saved. If a later store call fails, `fail` changes only the task record; the already-written JSON and Markdown still claim the review completed successfully.

Consumers can consequently observe contradictory audit states after a database or context failure. Publish the completed artifacts only after the required store operations succeed, or have the failure path remove or rewrite them, and cover this with a failing fake store.

<details><summary>中文</summary>

**不要在持久化成功前发布完成报告**

报告在 findings、审计决策、产物和报告元数据写库之前，就以 `StatusCompleted` 状态落盘。后续任一存储操作失败时，`fail` 只会更新任务记录，已经生成的 JSON 和 Markdown 仍会显示评审成功完成。

数据库或上下文异常后，报告消费者会看到相互矛盾的审计状态。建议在必要的存储操作成功后再发布完成态产物，或者在失败路径中删除或重写产物，并使用会注入失败的 fake store 覆盖该场景。

</details>

File: examples/code_review_agent/store/store.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+	// SavePermissionDecisions stores command governance decisions.
+	SavePermissionDecisions(ctx context.Context, taskID string, decisions []review.PermissionDecision) error
+	// SaveFilterDecisions stores noise-control filter decisions.
+	SaveFilterDecisions(ctx context.Context, taskID string, decisions []review.FilterDecision) error
+	// SaveArtifacts stores artifacts produced by the review.
+	SaveArtifacts(ctx context.Context, taskID string, artifacts []review.Artifact) error
+	// SaveReport stores the final report metadata and summary.
+	SaveReport(ctx context.Context, taskID string, report review.ReviewReport, jsonPath, markdownPath string) error
+	// CountFindings returns the number of stored findings.
+	CountFindings(ctx context.Context, taskID string) (int, error)
```
Comment (Rememorio) [inline examples/code_review_agent/store/store.go]: **P2: Keep the test-only counter out of Store**

`CountFindings` is only called by `sqlite_test.go`, and the concrete method explicitly describes itself as being for tests and demos. Including it in the advertised swappable `Store` contract forces every alternate backend to implement an operation the review pipeline never consumes.

Remove it from the interface; the concrete SQLite helper can remain if useful, or the test can assert `len(GetTask(...).Findings)`. This keeps the new public persistence contract consumer-oriented before it becomes a compatibility commitment.

<details><summary>中文</summary>

**不要把测试计数方法放入 Store 接口**

`CountFindings` 只在 `sqlite_test.go` 中被调用，具体实现的注释也明确说明它用于测试和演示。把它加入号称可替换的 `Store` 契约，会迫使每个后端实现一项评审主链路从不使用的能力。

建议将其从接口移除；如有需要，可保留 SQLite 具体类型上的辅助方法，或者让测试通过 `GetTask` 返回的 findings 数量断言。这样能在 API 固化前保持持久化接口以真实消费者需求为中心。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 4 inline comments below. The comments focus on issues worth addressing before merge.

Fresh review found four high-confidence issues: multiword quoted secrets can cross the model boundary, path-bearing fields remain unredacted in JSON/SQLite, completed reports are published before persistence succeeds, and the swappable Store interface includes a test-only method. Of seven previous comments, six are fixed; the offline dependency change is only partial because the skill runner's duplicate Go-check path still lacks the new skip/vendor policy.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 4 条行内评论。评论聚焦在合并前值得处理的问题。

本轮审查发现 4 个高置信度问题：带空格的引号密钥仍可能越过模型边界，路径字段在 JSON/SQLite 中未完整脱敏，数据库持久化成功前就会发布 completed 报告，以及可替换的 Store 接口包含仅供测试使用的方法。此前 7 条评论中有 6 条已修复；离线依赖问题仅部分修复，因为 Skill runner 的重复 Go 检查链路尚未应用新的跳过/vendor 策略。

</details>

**Comment by liuzengh** [comment]
感谢你对本 Issue 的贡献！

犀牛鸟活动的 Issue 提交阶段已经结束，本 PR 将关闭。再次感谢你的参与与付出！
